src/Nethermind/Nethermind.Specs/ChainSpecStyle/ChainSpecBasedSpecProvider.cs

```diff
@@ -41,162 +42,180 @@ private void BuildTransitions()
-                }
-                else if (propertyInfo.PropertyType == typeof(ulong?))
+                if (value is not null)
                 {
-                    var optionalTransition = (ulong?)propertyInfo.GetValue(propertyInfo.DeclaringType == typeof(ChainSpec) ? _chainSpec : propertyInfo.DeclaringType == typeof(EthashParameters) ? (object)_chainSpec.Ethash : _chainSpec.Parameters);
-                    if (optionalTransition is not null && optionalTransition.Value > (_chainSpec?.Genesis?.Timestamp ?? 0))
+                    IEnumerable<PropertyInfo> properties = value.GetType()
+                        .GetProperties(BindingFlags.Public | BindingFlags.Instance);
+
+                    foreach (PropertyInfo propertyInfo in properties)
                     {
-                        transitionTimestamps.Add(optionalTransition.Value);
+                        if (matchPropertyName(propertyInfo.Name))
+                        {
+                            if (propertyInfo.PropertyType == typeof(T))
+                            {
+                                Add(transitions, (T)propertyInfo.GetValue(value)!, minValueExclusive);
+                            }
+                            else if (propertyInfo.PropertyType == typeof(T?))
+                            {
+                                T? optionalTransition = (T?)propertyInfo.GetValue(value);
+                                if (optionalTransition is not null)
+                                {
+                                    Add(transitions, optionalTransition.Value, minValueExclusive);
+                                }
+                            }
+                        }
                     }
                 }
             }
 
-
             foreach (KeyValuePair<long, long> bombDelay in _chainSpec.Ethash?.DifficultyBombDelays ?? Enumerable.Empty<KeyValuePair<long, long>>())
             {
                 transitionBlockNumbers.Add(bombDelay.Key);
             }
-            _biggestBlockTransition = transitionBlockNumbers.Last();
-            TransitionActivations = transitionBlockNumbers.Skip(1).Select(bn => new ForkActivation(bn))
-                .Union(
-                transitionTimestamps.Select(ts => new ForkActivation(_biggestBlockTransition, ts))
-                )
-                .ToArray();
-            _transitions = new (ForkActivation, ReleaseSpec Release)[transitionBlockNumbers.Count + transitionTimestamps.Count];
+
+            TransitionActivations = CreateTransitionActivations(transitionBlockNumbers, transitionTimestamps);
+            _transitions = CreateTransitions(_chainSpec, transitionBlockNumbers, transitionTimestamps);
+            _firstTimestampActivation = TransitionActivations.FirstOrDefault(t => t.Timestamp is not null);
+
+            if (_chainSpec.Parameters.TerminalPowBlockNumber is not null)
+            {
+                MergeBlockNumber = (ForkActivation)(_chainSpec.Parameters.TerminalPowBlockNumber + 1);
```

Comment:
`_chainSpec.Parameters.TerminalPowBlockNumber` this is not added to the current chainspec

Reply:
I just moved this line.